### ISSUE TITLE
Support loading symbols for shared libraries from the trace

### ISSUE THREAD

**Issue by rocallahan** [mozilla/rr#2090]
Currently gdb tries to load symbols for shared libraries by grabbing the file names from tracee memory and trying to read those files in the local filesystem. That won't work well when the local files have changed or are simply not there. In those cases gdb needs to read symbols from files in our trace.

In 6b272d5a27a73c071ed492db0e4a2ccfcb04536f we added support for `vFile` in the remote protocol, and gdb would normally use this to load files from the remote target, but we disabled that by forcing `set sysroot /`. If we remove that command then gdb will request files from rr, so in principle we could return the contents of files from our trace, but there are a couple of problems:
* Shoveling large files through the remote protocol socket is slow. I don't think there's much we can do about this without modifying gdb.
* The filenames in tracee memory (and thus supplied by gdb) often refer to symbolic links to the real files. The "original file names" in the rr trace are obtained from `/proc/.../maps`, which gives the real names. Thus, identifying the correct trace file for the name provided by gdb is difficult.

**Comment by rocallahan** [comment]
For the second problem, one approach would be, after every successful `open` (or some subset of `open`s), if the file name argument is different from the path associated with the file descriptor (obtained via `/proc/.../fd/N`), record the pair in a "symlink database" associated with the trace. Then when gdb passes us a file name we can consult that database to resolve symlinks. Unfortunately `open` generally needs to be buffered for performance so doing this would be difficult.

For the first problem, `rr replay` could potentially create a temporary directory to serve as a "fake sysroot", populated with links mapping gdb file names to files in the trace directory.

**Comment by Keno** [comment]
Is the symlink thing actually a problem? I would assume gdb would also try to look at `/proc/.../maps` to find the file name of any mapped library (since the debugger rendez-vous structure only has the segment base).

**Comment by rocallahan** [comment]
I guess that's possible, but I don't see any code for it in gdb.

**Comment by rocallahan** [comment]
No, I don't.

**Comment by Keno** [comment]
Alright, I might give take a shot at implementing this in the morning then to see what happens.

**Comment by Keno** [comment]
You're correct that the symlinks are a problem. I hacked some heuristics together locally that worked fairly well which then allowed to properly debug the trace (branch here - https://github.com/Keno/rr/tree/kf/traceserve - very WIP - not cleaned up). To create the symlink database, we could have a special rr mode that does a replay, but sets a breakpoint on every `open`/`close` (even ones that go through the syscall buffer) to build a table of fd->opened filename, allowing us to associated those names to an eventual backing file if we mmap the fd.

**Comment by SemMulder** [comment]
Just leaving a comment to say this would be very useful for me! I'm currently debugging some Open CL kernels with the AMD APP SDK which unfortunately removes the .so files after completion. For now I just copy the the .so file from the results directory back to the right place. But this becomes quite tedious after a while ;).
